qiskit_optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,345 @@
+            raise QiskitOptimizationError(
+                f"Epsilon for warm-start QAOA needs to be between 0 and 0.5, "
+                f"actual value: {epsilon}")
+        self._epsilon = epsilon
+
+    def create_initial_variables(self, solution: np.ndarray) -> List[float]:
+        """
+        Creates initial variable values to warm start QAOA.
+
+        Args:
+            solution: a solution obtained for the relaxed problem.
+
+        Returns:
+            A list of initial variables constructed from a relaxed solution.
+        """
+        initial_variables = []
+
+        for variable in solution:
+            if variable < self._epsilon:
+                initial_variables.append(self._epsilon)
+            elif variable > 1. - self._epsilon:
+                initial_variables.append(1. - self._epsilon)
+            else:
+                initial_variables.append(variable)
+
+        return initial_variables
+
+    def create_initial_state(self, initial_variables: List[float]) -> QuantumCircuit:
+        """
+        Creates an initial state quantum circuit to warm start QAOA.
+
+        Args:
+            initial_variables: Already created initial variables.
+
+        Returns:
+            A quantum circuit that represents initial state.
+        """
+        circuit = QuantumCircuit(len(initial_variables))
+
+        for index, relaxed_value in enumerate(initial_variables):
+            theta = 2 * np.arcsin(np.sqrt(relaxed_value))
+            circuit.ry(theta, index)
+
+        return circuit
+
+    def create_mixer(self, initial_variables: List[float]) -> QuantumCircuit:
+        """
+        Creates an evolved mixer circuit as Ry(theta)Rz(-2beta)Ry(-theta).
+
+        Args:
+            initial_variables: Already created initial variables.
+
+        Returns:
+            A quantum circuit to be used as a mixer in QAOA.
+        """
+        circuit = QuantumCircuit(len(initial_variables))
+        beta = Parameter("beta")
+
+        for index, relaxed_value in enumerate(initial_variables):
+            theta = 2 * np.arcsin(np.sqrt(relaxed_value))
+
+            circuit.ry(-theta, index)
+            circuit.rz(-2.0 * beta, index)
+            circuit.ry(theta, index)
+
+        return circuit
+
+
+class WarmStartQAOAOptimizer(MinimumEigenOptimizer):
+    """A meta-algorithm that uses a pre-solver to solve a relaxed version of the problem.
+    Users must implement their own pre solvers by inheriting from the base class.
+
+    References:
+
+        [1]: Daniel J. Egger et al., Warm-starting quantum optimization.
+            `arXiv:2009.10095 <https://arxiv.org/abs/2009.10095>`_
+
+    """
+
+    def __init__(self,
+                 pre_solver: OptimizationAlgorithm,
+                 relax_for_pre_solver: bool,
+                 qaoa: QAOA,
+                 epsilon: float = 0.25,
+                 num_initial_solutions: int = 1,
+                 circuit_factory: Optional[WarmStartQAOAFactory] = None,
```

Comment:
Given the class name changed would a change of parameter name here help too?

Reply:
This is a reasonable question. Since `circuit_factory` sort of a prohibited name, I suggest `warming_factory`.

Reply:
Would 'factory' be simple enough to do the job - the type is right there for further info/

Reply:
I think `factory` is too generic, considering how often this term is used. So, I prefer have slightly more information in a variable name rather than just plain factory. This is why I suggest to have this name.

Reply:
How about `warm_start_factory`? The factory has does have more than just circuits, and maybe the meaning in the name here is closer to its use... Beyond that I am running out of ideas!

Reply:
OK, let it be `warm_start_factory`.